src/librustc/session/search_paths.rs

```diff
@@ -32,12 +28,16 @@ pub enum PathKind {
     All,
 }
 
-impl SearchPaths {
-    pub fn new() -> SearchPaths {
-        SearchPaths { paths: Vec::new() }
+impl PathKind {
+    pub fn matches(kind1: PathKind, kind2: PathKind) -> bool {
+        kind1 == PathKind::All ||
+        kind2 == PathKind::All ||
+        kind1 == kind2
     }
+}
 
-    pub fn add_path(&mut self, path: &str, output: config::ErrorOutputType) {
+impl SearchPath {
+    pub fn new(path: &str, output: config::ErrorOutputType) -> Self {
```

Comment:
I'd rename `new` to `from_cli_opt` and `new_all` to `from_sysroot` (and make the latter take the arguments of `make_target_lib_path`, and call that itself).

That's because `_new` seems to me like what I'd expect `SearchPath::new` to be, while the other two constructors are each more specific.